### PR TITLE
feat: add crosschain permission app

### PR DESCRIPTION
**Pull request by cosinlink**
### Description

Currently, Greenfield supports creating buckets and groups from the smart contracts deployed on BSC/opBNB. This means that the owner of a bucket can be a contract address. However, contract addresses cannot upload objects or put policies under the bucket, resulting in empty buckets owned by smart contracts.
To address this issue, we are going to introduce the cross-chain permission module. With this BEP, smart contracts can grant EOA addresses the ability to upload objects under their buckets, and even support additional functions.

### Rationale
In the current framework, users could only change resource permission on Greenfield. The proposal allows Greenfield users to change resource permission directly on the BSC/opBNB network.

### Changes

Notable changes: 
* add permission module
* ...

### CODE REVIEW

File: x/storage/keeper/cross_app_permission.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+				Status:    types.StatusFail,
+				Creator:   createPolicyPackage.Operator,
+				ExtraData: createPolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: err,
+		}
+	}
+
+	var policy permtypes.Policy
+	err = json.Unmarshal(createPolicyPackage.Data, &policy)
```
Comment (yutianwu) [inline x/storage/keeper/cross_app_permission.go:190]: why use json instead of abi

Comment (cosinlink) [reply x/storage/keeper/cross_app_permission.go:190]: 1. This data is passed in by the user from the frontend/client not from solidity contracts
2. This struct natively defines the serialization of json, using other methods requires additional code support
3. The additional overhead of using json serialization is not significant here

File: x/storage/keeper/cross_app.go [hunk 2]
```diff
@@ -22,4 +23,10 @@ func RegisterCrossApps(keeper Keeper) {
 	if err != nil {
 		panic(err)
 	}
+
+	permissionApp := NewPermissionApp(permissionKeeper)
```
Comment (alexgao001) [inline x/storage/keeper/cross_app.go:27]: maybe this `permissionApp` can be in `permission` module?

File: x/storage/keeper/cross_app_permission.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+func (app *PermissionApp) ExecuteAckPackage(ctx sdk.Context, appCtx *sdk.CrossChainAppContext, payload []byte) sdk.ExecuteResult {
+	pack, err := types.DeserializeCrossChainPackage(payload, types.PermissionChannelId, sdk.AckCrossChainPackageType)
+	if err != nil {
+		panic("deserialize Policy cross chain package error")
+	}
+
+	var operationType uint8
+	var result sdk.ExecuteResult
+	switch p := pack.(type) {
+	case *types.CreatePolicyAckPackage:
```
Comment (zjubfd) [inline x/storage/keeper/cross_app_permission.go:34]: Why we have Ack Packages?  Please explain in what case we may receive ACK Package???

File: x/storage/keeper/cross_app_permission.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func (app *PermissionApp) ExecuteFailAckPackage(ctx sdk.Context, appCtx *sdk.CrossChainAppContext, payload []byte) sdk.ExecuteResult {
+	pack, err := types.DeserializeCrossChainPackage(payload, types.PermissionChannelId, sdk.FailAckCrossChainPackageType)
+	if err != nil {
+		panic("deserialize Policy cross chain package error")
+	}
+
+	var operationType uint8
+	var result sdk.ExecuteResult
```
Comment (zjubfd) [inline x/storage/keeper/cross_app_permission.go:62]: So is the logic here, seems meaningless.

File: x/storage/keeper/cross_app_permission.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+		return sdk.ExecuteResult{
+			Payload: types.DeletePolicyAckPackage{
+				Status:    types.StatusFail,
+				ExtraData: deletePolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: types.ErrNoSuchPolicy,
+		}
+	}
+
+	_, err = app.permissionKeeper.DeletePolicy(ctx, policy.Principal, policy.ResourceType, policy.ResourceId)
```
Comment (zjubfd) [inline x/storage/keeper/cross_app_permission.go:148]: We need verify whether the `operator` is allowed to delete the policy.

File: x/storage/types/crosschain.go [hunk 6]
```diff
@@ -1286,3 +1386,114 @@ func (p UpdateGroupMemberAckPackage) MustSerialize() []byte {
...
+	}
+)
+
+func (p CreatePolicySynPackage) ValidateBasic() error {
+	if p.Operator.Empty() {
+		return sdkerrors.ErrInvalidAddress
+	}
+
+	var policy permtypes.Policy
+	err := json.Unmarshal(p.Data, &policy)
```
Comment (zjubfd) [inline x/storage/types/crosschain.go:1420]: why `json` is used? I hope rlp or even protobuf is accepted.

Comment (zjubfd) [reply x/storage/types/crosschain.go:1420]: once you get the policy, is any verification agains the policy?

File: x/storage/keeper/cross_app_permission.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+			Payload: types.CreatePolicyAckPackage{
+				Status:    types.StatusFail,
+				Creator:   createPolicyPackage.Operator,
+				ExtraData: createPolicyPackage.ExtraData,
+			}.MustSerialize(),
+			Err: err,
+		}
+	}
+
+	PolicyId, err := app.permissionKeeper.PutPolicy(ctx, &policy)
```
Comment (zjubfd) [inline x/storage/keeper/cross_app_permission.go:202]: So anyone is allowed to put any policy?

File: x/storage/keeper/cross_app.go (hunk 2 above)
Comment (cosinlink) [reply x/storage/keeper/cross_app.go]: it also depends storageKeeper

File: x/storage/keeper/cross_app_permission.go (hunk 3 above)
Comment (cosinlink) [reply x/storage/keeper/cross_app_permission.go]: removed

File: x/storage/keeper/cross_app_permission.go (hunk 5 above)
Comment (cosinlink) [reply x/storage/keeper/cross_app_permission.go:126]: added check

File: x/storage/types/crosschain.go (hunk 6 above)
Comment (cosinlink) [reply x/storage/types/crosschain.go:1420]: > once you get the policy, is any verification agains the policy?

added verification

File: x/storage/keeper/cross_app_permission.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+		Payload: types.CreatePolicyAckPackage{
+			Status:    types.StatusSuccess,
+			Id:        PolicyId.BigInt(),
+			Creator:   createPolicyPackage.Operator,
+			ExtraData: createPolicyPackage.ExtraData,
+		}.MustSerialize(),
+	}
+}
+
+func (app *PermissionApp) getResourceOwner(ctx sdk.Context, policy *permtypes.Policy) (resOwner sdk.AccAddress, err error) {
```
Comment (zjubfd) [inline x/storage/keeper/cross_app_permission.go:224]: can `getResourceOwner`, `normalizePrincipal`,`validatePrincipal` reuse the function in the  previous `permission` module?  
Reuse code can keep consistent with the previous logic.

**Review by zjubfd — APPROVED** [review]

**Review by alexgao001 — APPROVED** [review]

File: x/storage/keeper/cross_app_permission.go (hunk 1 above)
Comment (cosinlink) [reply x/storage/keeper/cross_app_permission.go]: modified To protobuf

**Review by Fynnss — APPROVED** [review]

**Pull request merged**